icu4j/main/classes/core/src/com/ibm/icu/util/LocalePriorityList.java

```diff
@@ -249,27 +270,48 @@ public LocalePriorityList build() {
          * @stable ICU 4.4
          */
         public LocalePriorityList build(boolean preserveWeights) {
-            // Walk through the input list, collecting the items with the same weights.
-            final Map<Double, Set<ULocale>> doubleCheck = new TreeMap<>(
-                    myDescendingDouble);
-            for (final ULocale lang : languageToWeight.keySet()) {
-                Double weight = languageToWeight.get(lang);
-                Set<ULocale> s = doubleCheck.get(weight);
-                if (s == null) {
-                    doubleCheck.put(weight, s = new LinkedHashSet<>());
-                }
-                s.add(lang);
+            if (built != null) {
+                // Calling build() again without changing anything in between.
+                // Just return the same immutable list.
+                return built;
             }
-            // We now have a bunch of items sorted by weight, then chronologically.
-            // We can now create a list in the right order
-            final Map<ULocale, Double> temp = new LinkedHashMap<>();
-            for (Entry<Double, Set<ULocale>> langEntry : doubleCheck.entrySet()) {
-                final Double weight = langEntry.getKey();
-                for (final ULocale lang : langEntry.getValue()) {
-                    temp.put(lang, preserveWeights ? weight : D1);
+            Map<ULocale, Double> temp;
+            if (hasWeights) {
+                // Walk through the input list, collecting the items with the same weights.
+                final TreeMap<Double, List<ULocale>> weightToLanguages =
```

Comment:
I don't know of a better way to sort these. What would that look like?

Note that in Java it's very common to use a TreeSet/TreeMap for sorting, and in this case we invert from a locale-to-weight map (insertion order) to a weight-to-locale map (sorted by weights).

I guess I could do here what I do in C++: Build an ArrayList of (locale, weight) pairs and sort by weight, but I don't think that's "better" in Java.

Reply:
I was thinking of the ArrayList approach, but as you pointed out, that would still require the object thrashing because you'd have to construct a new (locale, weight) pair for each entry.  That being said, TreeMap might be even worse than that in terms of object thrashing.  What you have is reasonably simple though, and I think ArrayList makes it a little less readable, so I'll just resolve this comment.